Add support for defineTensor

rdspring1 opened the pull request:
This PR creates the bindings for `define_tensor` and `DataType` enum bindings.
It corresponds with [Step 1: Create python bindings for basic Fusion IR](https://docs.google.com/document/d/1ftdNKu952EFmLANa36g0IrVaqAJ0eLKzgJESkQFnuVI/edit?tab=t.0#heading=h.e2plo2gt0gnb).

PR Stack:
#4403 Create IterDomain, TensorDomain, and TensorView bindings
#4406 Add support for define_tensor **<<< This PR.**
#4407 Add binding for Fusion IrContainer 
#4408 Create bindings for FusionDefinition and Add Binary op
#4409 Create python FusionDefinition for nvfuser_next

rdspring1 commented:
!test

rdspring1 commented:
!test

rdspring1 commented:
!test

rdspring1 commented on python/python_next/python_utils.h:
```diff
@@ -0,0 +1,80 @@
+
+namespace nvfuser::python {
+
+// [ Note stride order and contiguity vector ]
```
Duplicated utilities from `python_frontend`. I suppose I could make a common cpp folder between `python_frontend` and `python_next`. @jjsjann123

rdspring1 replied:
I will move these utilities to nvfuser `csrc/`.

rdspring1 replied:
I created `python/python_common/python_utils.h` to hold the helper functions shared between `python/python_direct` and `python/python_frontend`.

jacobhinkle reviewed (commented):
LGTM, but I'll let @jjsjann123 stamp so he can weigh in on temporary code duplication.

rdspring1 commented on python/python_next/ir.cpp:
```diff
@@ -121,12 +122,278 @@ IterDomain
+    const std::vector<int64_t>& stride_order) {
+  size_t rank = shape.size();
+  std::vector<bool> is_expand(rank);
+  for (size_t index : std::views::iota(0UL, rank)) {
```
This was pulled from #4077. I had implemented `arange` yet, so it still used `iota`.

rdspring1 commented on python/python_next/ir.cpp:
```diff
@@ -121,12 +122,278 @@ IterDomain
+            // Translate to TensorViewBuilder's view of the world.
+            std::vector<int64_t> dim_sizes;
+            dim_sizes.reserve(sizes.size());
+            for (size_t i : std::views::iota(0UL, sizes.size())) {
```
This was pulled from #4077. I had implemented `arange` yet, so it still used `iota`.

rdspring1 commented on python/python_next/ir.cpp:
```diff
@@ -121,12 +122,278 @@ IterDomain
+    // correctly with `contig_index` and `index`.
+    //
+    // stride_order[i] indicates that:
+    //   `logical_domain[i]` (and therefore `root_domain[i]` for input) maps
```
IDK, it is like this in https://github.com/NVIDIA/Fuser/blob/main/python/python_frontend/fusion_record.h#L1335-L1337.

rdspring1 replied:
I removed that part.

rdspring1 commented:
!test

rdspring1 commented:
!test

rdspring1 commented:
I created `python/python_common/python_utils.h` to hold the helper functions shared between `python/python_direct` and `python/python_frontend`.

From `define_tensor` in `python/python_frontend/python_bindings.cpp`, I extracted:
1. computeTensorDescriptor [Moved]
2. verifyShape [Moved]
3. getContiguityVec [New]
4. getTensorViewBuilderSizes [New]

From `TensorRecord` in `python/python_frontend/fusion_record.h`, I created:
1. `normalizeStrideOrder` --- Pulled from constructor; Updated to use `wrapDim` in `csrc/utils.h`.
2 `getExpanded` --- Pulled from `operator()`.

cc: @jacobhinkle @jjsjann123

jjsjann123 commented on CMakeLists.txt:
```diff
@@ -563,6 +566,7 @@ if(BUILD_PYTHON)
   list(APPEND NVFUSER_PYTHON_DIRECT_SRCS
     ${NVFUSER_PYTHON_DIRECT_BINDINGS}/extension.cpp
     ${NVFUSER_PYTHON_DIRECT_BINDINGS}/bindings.cpp
+    ${NVFUSER_PYTHON_DIRECT_BINDINGS}/enum.cpp
```
looks like we are not adding `python_utils.cpp` here?

jjsjann123 commented on python/python_direct/ir.cpp:
```diff
@@ -6,6 +6,7 @@
  */
 // clang-format on
 #include <bindings.h>
+#include <python_utils.h>
```
But we did add `python_utils.h` here? Felt like something is missing.

jjsjann123 commented on python/python_direct/ir.cpp:
```diff
@@ -121,12 +122,140 @@ IterDomain
+            std::vector<std::optional<bool>> contiguity;
+            std::vector<int64_t> stride_order;
+            std::tie(contiguity, stride_order) =
+                computeTensorDescriptor(sizes, strides);
```
this should be an undefined symbol?!

jjsjann123 commented on python/python_common/python_utils.cpp:
```diff
@@ -0,0 +1,236 @@
+ */
+// clang-format on
+
+#include <python_common/python_utils.h>
```
```suggestion
#include <python_utils.h>
```

rdspring1 replied:
```
  /opt/pytorch/nvfuser/python/python_common/python_utils.cpp:9:10: fatal error: python_utils.h: No such file or directory
      9 | #include <python_utils.h>
        |          ^~~~~~~~~~~~~~~~
```

rdspring1 replied:
I added `target_include_directories(codegen_internal PUBLIC ${NVFUSER_PYTHON_COMMON})`

jjsjann123 commented on python/python_frontend/fusion_record.h:
```diff
@@ -13,6 +13,7 @@
 #include <ir/interface_nodes.h>
 #include <ops/all_ops.h>
 #include <options.h>
+#include <python_common/python_utils.h>
```
```suggestion
#include <python_utils.h>
```

jjsjann123 reviewed (commented):
I think I missed something obvious, since python direct library didn't include the new util file as its source, I'm surprised that we didn't get an undefined symbol at linking.

rdspring1 commented:
- `python/python_common/python_utils.h` is used in `python/python_frontend/fusion_record.h`.
- `python/python_frontend/fusion_record.h` is used in `csrc/serde/fusion_record.h`
- Since `python/python_common/python_utils.h` is used in `csrc/`, it is compiled once with `codegen_internals` target at https://github.com/NVIDIA/Fuser/pull/4406/files#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR311.
- Both `python_frontend` and `python_direct` targets link against `codegen_internals`.

rdspring1 commented:
!test

rdspring1 commented:
!test

jjsjann123 commented:
> * Since `python/python_common/python_utils.h` is used in `csrc/`, it is compiled once with `codegen_internals` target at https://github.com/NVIDIA/Fuser/pull/4406/files#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR311.
> * Both `python_frontend` and `python_direct` targets link against `codegen_internals`.

ah, sorry I mistakenly thought the `python_common/` stuff was added to `NVFUSER_PYTHON_SRCS`, instead of the `NVFUSER_SRCS`. I should have been more careful.

Having said that, it felt a bit strange to leave them in the python folder, if they are not built by the pybind stuff.

jjsjann123 commented on python/python_common/python_utils.cpp:
```diff
@@ -0,0 +1,236 @@
+    bool static_sizes) {
+  // TensorViewBuilder assumes any dim with a compile time constant
+  // size == 1 is a broadcast axis, symbolic sizes are
+  // identified by -1, and size == 0 is not supported.
```
nitpick: the comment here mentioned about `size == 0` is not supported, yet we do allow 0 in `sizes` in the logic below.

jjsjann123 reviewed (approved):
nitpick on the location of python_utils.h/cpp.
But overall changes looks good to me.

Thanks a lot for addressing the review comments.

rdspring1 commented:
!test

rdspring1 commented:
I mean we have a lot of `python_frontend` cpp files built with the `codegen_internals` target. It seems unusual to me.

```cmake
 26 if(BUILD_PYTHON)
 25   list(APPEND NVFUSER_SRCS
 24     ${NVFUSER_PYTHON_BINDINGS}/distributed_tensor.cpp
 23     ${NVFUSER_PYTHON_BINDINGS}/fusion_cache.cpp
 22     ${NVFUSER_PYTHON_BINDINGS}/fusion_definition.cpp
 21     ${NVFUSER_PYTHON_BINDINGS}/fusion_state.cpp
 20     ${NVFUSER_PYTHON_BINDINGS}/segmentation.cpp
 19     ${NVFUSER_PYTHON_BINDINGS}/translation.cpp
 18     ${NVFUSER_PYTHON_BINDINGS}/translation_utils.cpp
 17     ${NVFUSER_SRCS_DIR}/serde/fusion_record.cpp
 16     ${NVFUSER_PYTHON_COMMON}/python_utils.cpp
 15   )
 14 endif()
```
